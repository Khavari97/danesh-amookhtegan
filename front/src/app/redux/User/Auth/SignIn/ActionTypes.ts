export const REQUEST = 'REQUEST';
export let FETCH_SUCCESS = 'FETCH_SUCCESS';
export let FETCH_FAIL = 'FETCH_FAIL';

export interface IRequest {
    type: typeof REQUEST,
}

export interface ISuccess {
    type: typeof FETCH_SUCCESS,
    payload: string
}

export interface IFail {
    type: typeof FETCH_FAIL,
    payload: string
}


export interface IState {
    loading: boolean,
    res: string,
    error: string
}

export const INITIAL_STATE: IState = {
    loading: true,
    res: "",
    error: ""
}
export type FetchDispatchTypes = IFail | ISuccess | IRequest