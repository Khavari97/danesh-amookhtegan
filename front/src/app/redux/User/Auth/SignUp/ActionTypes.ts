export const SET_PASSWORD = 'SET_PASSWORD';
export const SET_PROFILE = 'SET_PROFILE';

export interface IState {
    password: string,
    profile: object,
}

export interface IPassword {
    type: typeof SET_PASSWORD;
    payload: string;
}

export interface IProfile {
    type: typeof SET_PROFILE;
    payload: object;
}

export const INITIAL_STATE: IState = {
    password: '',
    profile: {},
}

export type SignUpTypes = IPassword | IProfile