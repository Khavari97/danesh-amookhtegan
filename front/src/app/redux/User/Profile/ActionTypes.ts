export const SET_PROVINCES = 'SET_PROVINCES';
export const SET_CITIES = 'SET_CITIES';
export const SET_UNIVERSITY = 'SET_UNIVERSITY'
export const SET_DECRIPTION = 'SET_DECRIPTION'
export const GET_PROFILE = 'GET_PROFILE'
export const GET_PARENT_SKILLS = 'GET_PARENT_SKILLS'
export const GET_SKILLS = 'GET_SKILLS'
export const GET_RESUME = 'GET_RESUME'
export const GET_AVATAR = 'GET_AVATAR'
export const GET_BRIEF_PROFILE = 'GET_BRIEF_PROFILE'

interface University {
    id: number | any
    code: string,
    name: string,
    city: number
    province?: province | any
}

interface Universities extends Array<University> {
}

export interface IUniversities {
    type: typeof SET_UNIVERSITY,
    payload: Universities | any
}

interface City {
    id: number | any
    code: string,
    name: string,
    province?: province | any
}

interface Cities extends Array<City> {
}

export interface ICities {
    type: typeof SET_CITIES,
    payload: Cities | any
}

interface province {
    id: number | any
    code: string,
    name: string
}

interface provinces extends Array<province> {
}

export interface IProvinces {
    type: typeof SET_PROVINCES,
    payload: provinces | any
}


export interface IProfile {
    type: typeof SET_PROVINCES,
    payload: provinces | any
}

export interface IDescription {
    type: typeof SET_DECRIPTION,
    payload: string
}

//profile get
export interface IGetProfile {
    loading: boolean,
    type: typeof GET_PROFILE,
    description: string | any
    data: any
}

export interface IGET_BRIEF_PROFILE {
    type: typeof GET_BRIEF_PROFILE,
    briefprofile: any
    avatar: any
}

export interface IGetResume {
    type: typeof GET_RESUME,
    resume: string
}

export interface IGetAvatar {
    type: typeof GET_AVATAR,
    avatar: string
}

export interface ParentSkill {
    code: string
    name: string
    skill: any
    image: string | any
    isAdd?: boolean
}

interface ParentSkills extends Array<ParentSkill> {
}

export interface IGetParentSkills {
    type: typeof GET_PARENT_SKILLS,
    parentSkills: ParentSkills
}

export interface Skill {
    id: number
    name: string
}

interface Skills extends Array<Skill> {
}

export interface IGetSkills {
    type: typeof GET_SKILLS,
    skills: Skills
}

export interface IState {
    provinces: provinces | any
    profile: object
    cities: Cities
    universities: Universities
    description: string | any
    data: object | any
    parentSkills: ParentSkills
    skills: Skills
    loading: boolean
    resume: string | null
    avatar: any
    briefprofile: any
}

export const INITIAL_STATE: IState = {
    loading: true,
    provinces: [],
    cities: [],
    universities: [],
    profile: {},
    description: '',
    data: {},
    parentSkills: [],
    skills: [],
    resume: null,
    avatar: null,
    briefprofile: {}
}

export type ProfileTypes = IProvinces | ICities | IUniversities | IDescription |
    IGetProfile | IGetParentSkills | IGetSkills | IGetResume | IGetAvatar | IGET_BRIEF_PROFILE
