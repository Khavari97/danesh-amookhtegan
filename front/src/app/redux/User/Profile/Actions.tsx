import {
    GET_AVATAR,
    GET_BRIEF_PROFILE,
    GET_PARENT_SKILLS,
    GET_PROFILE,
    GET_SKILLS,
    SET_CITIES,
    SET_PROVINCES,
    SET_UNIVERSITY
} from "./ActionTypes";
import {Dispatch} from "redux";
import {apiClient, defaultConfig} from "../../../api/_api";
import Snackbar from "./../../../components/material-ui/Snackbar/SnackbarUtils";


export const GetBriefProfile = () => (
    dispatch: Dispatch
) => {
    try {
        return apiClient
            .get("user/brief-profile/", {
                consoleMessage: true,
            })
            .then((response) => {
                dispatch({
                    type: GET_BRIEF_PROFILE,
                    briefprofile: response,
                    avatar: response.avatar
                });
            })
            .catch((error) => {
            });
    } catch {
    }
}


export const GetAvatar = () => (
    dispatch: Dispatch
) => {
    try {
        return apiClient
            .get("user/profile/avatar/", {
                consoleMessage: true,
                headers: {...defaultConfig.headers, 'Content-Type': 'multipart/form-data'}
            })
            .then((response) => {
                dispatch({
                    type: GET_AVATAR,
                    avatar: response.avatar,
                });

            })
            .catch((error) => {
            });
    } catch {
    }
}


export const FilterCity = (code: number | string) => (
    dispatch: Dispatch
) => {
    try {
        return apiClient
            .get("list/cities/?province=" + code, {
                consoleMessage: true,
            })
            .then((response) => {
                dispatch({
                    type: SET_CITIES,
                    payload: response,
                });
            })
            .catch((error) => {
            });
    } catch {
    }
}

export const GetSkills = () =>  (
    dispatch: Dispatch
) => {
    try {
        return apiClient.get("user/profile/skills/").then((response: any) => {
            dispatch({
                type: GET_SKILLS,
                skills: response.skills
            });
        })
            .catch(() => {
                Snackbar.error("خطایی رخ داده لطفا دوباره امتحان کنید")
            })
    } catch {
    }
}


export const GetParentSkills = () => async (
    dispatch: Dispatch
) => {
    try {
        return apiClient.get('list/skills/').then(async (response: any) => {
            response.map((element: any, index: number) => {
                element.isAdd = false
            })
            dispatch({
                type: GET_PARENT_SKILLS,
                parentSkills: response,
            });
        })
            .catch(() => {
                Snackbar.error("خطایی رخ داده لطفا دوباره امتحان کنید")
            })
    } catch {


    }
}


export const GetProfile = () => async (
    dispatch: Dispatch
) => {
    try {
        await apiClient.get("user/profile/", {notifyMessage: true}).then((response: any) => {
            const c = 'description';
            const description = response.description
            const {[c]: _, ...withoutDescripe} = response;
            let data = {
                firstName: withoutDescripe.firstName,
                lastName: withoutDescripe.lastName,
                username: withoutDescripe.username,
                email: withoutDescripe.email,
                province: (withoutDescripe.city) ? (withoutDescripe.city.province) ? withoutDescripe.city.province.name : null : null,
                university: (withoutDescripe.university) ? withoutDescripe.university.name : null,
                city: (withoutDescripe.city) ? withoutDescripe.city.name : null,
                cityObject: (withoutDescripe.city) ? withoutDescripe.city.id : null,
                provinceObject: (withoutDescripe.city) ? (withoutDescripe.city.province) ? withoutDescripe.city.province.id : null : null,
                universityObject: (withoutDescripe.university) ? withoutDescripe.university.id : null,
                phoneNumber: withoutDescripe.phoneNumber,
                genderDisplay: withoutDescripe.genderDisplay,
                gender: (withoutDescripe.gender) ? withoutDescripe.gender : null,
            }
            dispatch({
                type: GET_PROFILE,
                data: data,
                description: description
            });
        })
    } catch {
    }
}

export const ProvinceApi = () => (
    dispatch: Dispatch
) => {

    try {
        apiClient
            .get("list/provinces/", {
                consoleMessage: true,
            })
            .then((response) => {
                dispatch({
                    type: SET_PROVINCES,
                    payload: response,
                });
            })

        apiClient
            .get("list/cities/", {
                consoleMessage: true,
            })
            .then((response) => {
                dispatch({
                    type: SET_CITIES,
                    payload: response,
                });
            })

        apiClient
            .get("list/universities/", {
                consoleMessage: true,
            })
            .then((response) => {
                dispatch({
                    type: SET_UNIVERSITY,
                    payload: response,
                });
            })
    } catch (e) {
    }
};
