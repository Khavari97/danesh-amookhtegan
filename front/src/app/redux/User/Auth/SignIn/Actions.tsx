import {FETCH_FAIL, FETCH_SUCCESS, FetchDispatchTypes, IRequest, ISuccess, REQUEST,} from "./ActionTypes";
import {Dispatch} from "redux";
import {apiClient} from "../../../../api/_api";
import Snackbar from "../../../../components/material-ui/Snackbar/SnackbarUtils";

export const Request = (): IRequest => {
    return {
        type: REQUEST,
    };
};

export const FetchSuccess = (data: string): ISuccess => {
    return {
        type: FETCH_SUCCESS,
        payload: data,
    };
};

// if this works we should add user and pass
export const Api = (username: string, password: string) => (
    dispatch: Dispatch<FetchDispatchTypes>
) => {
    try {

        dispatch({
            type: REQUEST,
        });

        return apiClient
            .post("api/token/", {
                data: {
                    username: username,
                    password: password,
                },
            })
            .then((response) => {
                dispatch({
                    type: FETCH_SUCCESS,
                    payload: response.access,
                });
                localStorage.setItem("access", response.access);
                window.location.href = "/"
            })
            .catch((error) => {
                dispatch({
                    type: FETCH_FAIL,
                    payload: error.data.detail,
                });
                Snackbar.error("رمز عبور یا نام کاربری اشتباه است.")
            });


    } catch (e) {
    }
};
