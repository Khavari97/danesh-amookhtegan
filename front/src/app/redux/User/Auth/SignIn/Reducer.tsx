import {FetchDispatchTypes, INITIAL_STATE, IState} from "./ActionTypes";


const ReducerSignIn = (state = INITIAL_STATE, action: FetchDispatchTypes): IState => {
    switch (action.type) {
        case 'REQUEST': {
            return {
                ...state,
                loading: true
            }
        }

        case 'FETCH_SUCCESS': {
            return {
                ...state,
                loading: false,
                res: action.payload
            }
        }
        case 'FETCH_FAIL': {
            return {
                ...state,
                loading: false,
                error: action.payload
            }
        }
        default:
            return state
    }

}


export default ReducerSignIn