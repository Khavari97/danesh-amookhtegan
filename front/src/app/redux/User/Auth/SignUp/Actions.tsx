import {IPassword, SET_PASSWORD} from "./ActionTypes";

export const FetchSuccess = (password: string): IPassword => {
    return {
        type: SET_PASSWORD,
        payload: password
    }
}