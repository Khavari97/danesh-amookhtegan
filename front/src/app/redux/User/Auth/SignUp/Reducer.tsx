import {INITIAL_STATE, IState, SET_PASSWORD, SET_PROFILE, SignUpTypes} from "./ActionTypes";


const ReducerSignup = (state = INITIAL_STATE, action: SignUpTypes): IState => {
    switch (action.type) {
        case SET_PASSWORD: {
            return {
                ...state,
                password: action.payload
            }
        }

        case SET_PROFILE: {
            return {
                ...state,
                profile: action.payload
            }
        }
        default:
            return state
    }

}


export default ReducerSignup