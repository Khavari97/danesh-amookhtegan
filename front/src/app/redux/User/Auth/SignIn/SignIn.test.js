import configureStore from 'redux-mock-store'
import { applyMiddleware, createStore } from 'redux'
import thunk from "redux-thunk";
import Reducer from "./Reducer";
import { actionsToStateSnapshot } from 'redux-state-snapshot'
import { FETCH_SUCCESS, FETCH_FAIL, REQUEST } from "./ActionTypes";
import moxios from 'moxios';
import { apiUrl } from '../../../../api/_api';
import { Api, FetchSuccess } from './Actions';

const mock_access_token = "1234567"
const mock_user_name = "atena"
const mock_password = "456"

const middlewares = [thunk]
const mockStore = configureStore(middlewares)


it('should dispatch action', () => {
    moxios.install();
    moxios.stubRequest(apiUrl + `api/token/`, {
        status: 200, response: {
            access: mock_access_token,
        },
    });
    // Initialize mockstore with empty state
    const createState = initialState => actions => actions.reduce(Reducer, initialState);
    const initialState = createState({
        loading: true,
        res: "",
        error: ""
    });
    const store = mockStore(initialState)
    const actions = store.getActions()


    return store.dispatch(Api(mock_user_name, mock_password)).then(() => {

        let dispatchedActions = actions
        let dispatchedTypes = dispatchedActions.map(action => action.type)
        expect(dispatchedTypes).toEqual([REQUEST,FETCH_SUCCESS])

        expect(actions[0]).toEqual({
            type: REQUEST,
        });
        const state = store.getState()
        expect(actions[1]).toEqual(FetchSuccess(mock_access_token));
        expect(state).toEqual({
            loading: false,
            res: mock_access_token,
            error: ""
        });
    });

})

// it('not dispatch action', () => {
//     moxios.install();
//     moxios.stubRequest(apiUrl + `api/token/`, {
//         status: 400,
//     });
//     // Initialize mockstore with empty state
//     const createState = initialState => actions => actions.reduce(Reducer, initialState);
//     const initialState = createState({
//         loading: true,
//         res: "",
//         error: ""
//     });
//     const store = mockStore(initialState)
//     const actions = store.getActions()


//     return store.dispatch(Api(mock_user_name, mock_password)).then(() => {
        
//         expect(actions[0]).toEqual({
//             type: REQUEST,
//         });
//         const state = store.getState()
 



//         expect(actions[1]).toEqual(FetchSuccess(mock_access_token));
//         expect(state).toEqual({
//             loading: false,
//             res: mock_access_token,
//             error: ""
//         });
//     });

// })

it('to snapshot on failure', () => {
    const actionStateSnapshot = actionsToStateSnapshot([thunk])
    actionStateSnapshot(() => fetchDataFailure(), Reducer)
})
it('to snapshot on success', () => {
    const actionStateSnapshot = actionsToStateSnapshot()
    actionStateSnapshot(() => FetchSuccess('test'), Reducer)
})