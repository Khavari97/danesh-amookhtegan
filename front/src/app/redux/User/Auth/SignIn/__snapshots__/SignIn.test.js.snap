// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`to snapshot on success: FETCH_SUCCESS 1`] = `
Object {
  "error": "",
  "loading": false,
  "res": "test",
}
`;
